gslib/daisy_chain_wrapper.py

```diff
@@ -239,34 +240,53 @@ def read(self, amt=None):  # pylint: disable=invalid-name
           'Invalid HTTP read size %s during daisy chain operation, '
           'expected <= %s.' % (amt, constants.TRANSFER_BUFFER_SIZE))
 
-    while True:
-      with self.lock:
-        if self.buffer:
-          break
-        if AcquireLockWithTimeout(self.download_exception_lock, 30):
-          if self.download_exception:
-            # Download thread died, so we will never recover. Raise the
-            # exception that killed it.
-            raise self.download_exception  # pylint: disable=raising-bad-type
-        else:
-          if not self.download_thread.is_alive():
-            raise Exception('Download thread died suddenly.')
-      # Buffer was empty, yield thread priority so the download thread can fill.
-      time.sleep(0)
+    if self._unused_data_from_previous_read:
+      valid_data = self._unused_data_from_previous_read[:amt]
+      self._unused_data_from_previous_read = (
+          self._unused_data_from_previous_read[amt:])
+    else:
+      valid_data = b''
+
+      while True:
+        with self.lock:
+          if self.buffer:
+            break
+          if AcquireLockWithTimeout(self.download_exception_lock, 30):
+            if self.download_exception:
+              # Download thread died, so we will never recover. Raise the
+              # exception that killed it.
+              raise self.download_exception  # pylint: disable=raising-bad-type
+          else:
+            if not self.download_thread.is_alive():
+              raise Exception('Download thread died suddenly.')
+        # Buffer was empty, yield thread priority so the download thread can fill.
+        time.sleep(0)
     with self.lock:
-      # TODO: Need to handle the caller requesting less than a
-      # transfer_buffer_size worth of data.
-      data = self.buffer.popleft()
+      # Note that we are not handling the rare case where len(valid_data) < amt
+      # and there are elements in the buffer.
+      # For e.g if valid_data = 'foo' and amt = 5, and we have 'barbaz' in
```

Comment:
e.g. = "For example", so writing "For e.g." = "For for example"
https://www.aje.com/arc/editing-tip-using-eg-and-ie/#:~:text=e.g.%20is%20the%20abbreviation%20for,implies%20that%20there%20are%20other

Reply:
Done